galter_subjects_utils/updater.py

```diff
@@ -269,36 +293,68 @@ def _update_rdm_records(self):
             )
 
     def _remove_rdm_subjects(self):
-        """Remove subjects from the Subjects entries."""
+        """Remove subjects from the Subjects entries.
+
+        We have to resort to low-level commands because the high-level ones
+        are not made for bulk operations. We've checked the implications
+        and we should be fine (at least at time of writing).
+        """
+        ids_for_removal = [
+            op["id"] for op in self._ops_data
+            if op.get("type") in ["remove", "replace"]
+        ]
         service = current_service_registry.get("subjects")
 
-        remove_ops = filter_ops_by_type(self._ops_data, "remove")
-        replace_ops = filter_ops_by_type(self._ops_data, "replace")
-        for op in remove_ops + replace_ops:
-            try:
-                service.delete(
-                    system_identity,
-                    op["id"]
-                )
-            except PersistentIdentifierError:
-                # For any persistent identifier related problem
-                # we just ignore it and make sure we can continue.
-                # Main scenario is when a subject has already been
-                # "service.delete"'d in which case its PID is only marked as
-                # deleted which interferes with service.delete and
-                # service.create.
-                pass
-
-            # Purge (completely delete) backing subject PID
-            # This operation is idempotent.
-            delete_pid_stmt = (
+        model_cls = service.record_cls.model_cls
+        size_of_batch = 200  # Maybe TODO: make configurable
```

Comment:
See #27 again.

Reply:
I think making it configurable is a smart move, but I understand if you're busy with other work it might not be a high priority. In my own experience dealing with benchmarking / profiling where I don't add some means to change these types of options I always wish I had. To be fair though most of those instances are not particularly complicated 😆

Reply:
That's fair. I will be running some benchmarking tonight and hope to have some good news...

Reply:
Profiling results for 620454 subject replacements. This is for overall update execution, and there weren't 241 records to update in this run, so it's a more optimistic scenario than normal. 

```
Before: 347.97 mins = 20878.20 secs 
After:   ~1.80 mins =   107.36 secs 
```

So a factor of 194 or 99.49% decrease in time. This is quite good, but it shows more how the original way we were doing it was very bad.

Reply:
> it shows more how the original way we were doing it was very bad
Live and learn I guess 🤷 I think it's natural to turn to the database level when dealing with these kinds of problems. In my databases class that I'm finishing up we had a practice test question about whether to deal with analysis of a large data set using Python or Postgres. The question specifically asked for arguing the Postgres side of things, but this is an interesting example of how our expectations don't always work out in practice.